File: src/osscodeiq/detectors/shell/powershell_detector.py
"""PowerShell script detector for functions, modules, parameters, and dot-sourcing."""

from __future__ import annotations

import re

from osscodeiq.detectors.base import DetectorContext, DetectorResult
from osscodeiq.detectors.utils import decode_text
from osscodeiq.models.graph import (
    EdgeKind,
    GraphEdge,
    GraphNode,
    NodeKind,
    SourceLocation,
)

_FUNC_RE = re.compile(r'function\s+([\w-]+)\s*(?:\([^)]*\))?\s*\{', re.IGNORECASE)
_IMPORT_RE = re.compile(r'Import-Module\s+(\S+)', re.IGNORECASE)
_DOT_SOURCE_RE = re.compile(r'\.\s+["\']?(\S+\.ps(?:1|m1))["\']?')
_PARAM_RE = re.compile(r'\[Parameter[^]]*\]\s*\[(\w+)\]\s*\$(\w+)')
_CMDLET_BINDING_RE = re.compile(r'\[CmdletBinding\(\)\]', re.IGNORECASE)


class PowerShellDetector:
    """Detects PowerShell script structures: functions, modules, parameters, and dot-sourcing."""

    name: str = "powershell"
    supported_languages: tuple[str, ...] = ("powershell",)

    def detect(self, ctx: DetectorContext) -> DetectorResult:
        result = DetectorResult()
        text = decode_text(ctx)
        lines = text.split("\n")

        # Track which functions have CmdletBinding for advanced function marking
        # We do a two-pass approach: first find functions, then check for CmdletBinding nearby
        func_positions: list[tuple[int, str]] = []

        # Detect function definitions
        for i, line in enumerate(lines):
            m = _FUNC_RE.search(line)
            if not m:
                continue
            func_name = m.group(1)
            func_positions.append((i, func_name))

            # Check if [CmdletBinding()] appears within the next few lines
            is_advanced = False
            for j in range(i + 1, min(i + 5, len(lines))):
                if _CMDLET_BINDING_RE.search(lines[j]):
                    is_advanced = True
                    break

            props: dict = {}
            if is_advanced:
                props["advanced_function"] = True

            result.nodes.append(GraphNode(
                id=f"{ctx.file_path}:{func_name}",
                kind=NodeKind.METHOD,
                label=func_name,
                fqn=func_name,
                module=ctx.module_name,
                location=SourceLocation(
                    file_path=ctx.file_path,
                    line_start=i + 1,
                ),
                properties=props,
            ))

        # Detect Import-Module statements
        for i, line in enumerate(lines):
            m = _IMPORT_RE.search(line)
            if not m:
                continue
            module_name = m.group(1)
            result.edges.append(GraphEdge(
                source=ctx.file_path,
                target=module_name,
                kind=EdgeKind.IMPORTS,
                label=f"{ctx.file_path} imports module {module_name}",
            ))

        # Detect dot-sourcing
        for i, line in enumerate(lines):
            m = _DOT_SOURCE_RE.search(line)
            if not m:
                continue
            sourced_file = m.group(1)
            result.edges.append(GraphEdge(
                source=ctx.file_path,
                target=sourced_file,
                kind=EdgeKind.IMPORTS,
                label=f"{ctx.file_path} dot-sources {sourced_file}",
            ))

        # Detect typed parameters with [Parameter] attribute
        for i, line in enumerate(lines):
            m = _PARAM_RE.search(line)
            if not m:
                continue
            param_type = m.group(1)
            param_name = m.group(2)

            result.nodes.append(GraphNode(
                id=f"{ctx.file_path}:param:{param_name}",
                kind=NodeKind.CONFIG_DEFINITION,
                label=f"${param_name}: {param_type}",
                fqn=param_name,
                module=ctx.module_name,
                location=SourceLocation(
                    file_path=ctx.file_path,
                    line_start=i + 1,
                ),
                properties={"param_type": param_type},
            ))

        return result

File: src/main/java/io/github/randomcodespace/iq/detector/script/shell/PowerShellDetector.java
package io.github.randomcodespace.iq.detector.script.shell;

import io.github.randomcodespace.iq.detector.AbstractRegexDetector;
import io.github.randomcodespace.iq.detector.DetectorContext;
import io.github.randomcodespace.iq.detector.DetectorResult;
import io.github.randomcodespace.iq.model.CodeEdge;
import io.github.randomcodespace.iq.model.CodeNode;
import io.github.randomcodespace.iq.model.EdgeKind;
import io.github.randomcodespace.iq.model.NodeKind;
import org.springframework.stereotype.Component;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import io.github.randomcodespace.iq.detector.DetectorInfo;

@DetectorInfo(
    name = "powershell",
    category = "config",
    description = "Detects PowerShell script structure (functions, modules, imports)",
    languages = {"powershell"},
    nodeKinds = {NodeKind.CONFIG_DEFINITION, NodeKind.METHOD, NodeKind.MODULE},
    edgeKinds = {EdgeKind.IMPORTS}
)
@Component
public class PowerShellDetector extends AbstractRegexDetector {

    private static final Pattern FUNC_RE = Pattern.compile("function\\s+([\\w-]+)\\s*(?:\\([^)]*\\))?\\s*\\{", Pattern.CASE_INSENSITIVE);
    private static final Pattern IMPORT_RE = Pattern.compile("Import-Module\\s+(\\S+)", Pattern.CASE_INSENSITIVE);
    private static final Pattern DOT_SOURCE_RE = Pattern.compile("\\.\\s+[\"']?(\\S+\\.ps(?:1|m1))[\"']?");
    private static final Pattern PARAM_RE = Pattern.compile("\\[Parameter[^]]*\\]\\s*\\[(\\w+)\\]\\s*\\$(\\w+)");
    private static final Pattern CMDLET_BINDING_RE = Pattern.compile("\\[CmdletBinding\\(\\)\\]", Pattern.CASE_INSENSITIVE);

    @Override
    public String getName() { return "powershell"; }

    @Override
    public Set<String> getSupportedLanguages() { return Set.of("powershell"); }

    @Override
    public DetectorResult detect(DetectorContext ctx) {
        String text = ctx.content();
        if (text == null || text.isEmpty()) return DetectorResult.empty();

        List<CodeNode> nodes = new ArrayList<>();
        List<CodeEdge> edges = new ArrayList<>();
        String fp = ctx.filePath();
        String[] lines = text.split("\n", -1);

        for (int i = 0; i < lines.length; i++) {
            Matcher m = FUNC_RE.matcher(lines[i]);
            if (m.find()) {
                String funcName = m.group(1);
                boolean isAdvanced = false;
                for (int j = i + 1; j < Math.min(i + 5, lines.length); j++) {
                    if (CMDLET_BINDING_RE.matcher(lines[j]).find()) { isAdvanced = true; break; }
                }
                CodeNode n = new CodeNode(); n.setId(fp + ":" + funcName);
                n.setKind(NodeKind.METHOD); n.setLabel(funcName); n.setFqn(funcName);
                n.setFilePath(fp); n.setLineStart(i + 1);
                if (isAdvanced) n.getProperties().put("advanced_function", true);
                nodes.add(n);
            }

            Matcher im = IMPORT_RE.matcher(lines[i]);
            if (im.find()) {
                CodeEdge e = new CodeEdge(); e.setId(fp + ":imports:" + im.group(1));
                e.setKind(EdgeKind.IMPORTS); e.setSourceId(fp);
                e.setTarget(new CodeNode(im.group(1), NodeKind.MODULE, im.group(1)));
                edges.add(e);
            }

            Matcher dm = DOT_SOURCE_RE.matcher(lines[i]);
            if (dm.find()) {
                CodeEdge e = new CodeEdge(); e.setId(fp + ":dotsource:" + dm.group(1));
                e.setKind(EdgeKind.IMPORTS); e.setSourceId(fp);
                e.setTarget(new CodeNode(dm.group(1), NodeKind.MODULE, dm.group(1)));
                edges.add(e);
            }

            Matcher pm = PARAM_RE.matcher(lines[i]);
            if (pm.find()) {
                String paramType = pm.group(1);
                String paramName = pm.group(2);
                CodeNode n = new CodeNode(); n.setId(fp + ":param:" + paramName);
                n.setKind(NodeKind.CONFIG_DEFINITION);
                n.setLabel("$" + paramName + ": " + paramType);
                n.setFqn(paramName); n.setFilePath(fp); n.setLineStart(i + 1);
                n.getProperties().put("param_type", paramType);
                nodes.add(n);
            }
        }

        return DetectorResult.of(nodes, edges);
    }
}
